### PR TITLE
Server manager

### PR DESCRIPTION
**Pull request by dVaGaymer**

### CODE REVIEW

File: src/class/ServerManager/ServerManager.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+	void start();
+	void terminate();
+	ServerManager::status setup();
+
+  private:
+	Poll _poll;
+	const IConf *_conf;
+	enum status _status;
+	std::vector<Server> _virtualServers;
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.hpp:43]: `_virtualServers -> _virtual_servers`

File: src/class/ServerManager/ServerManager.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
+
+#include <ServerManager/ServerManager.hpp>
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:2]: func should be in camelCase

these func are not part of a namespace or a class

File: src/class/ServerManager/ServerManager.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	for (; it != end; it++)
+	{
+		try
+		{
+			it->bindSockets();
+			it->listenSockets();
+		}
+		catch (std::exception &e)
+		{
+			return ServerManager::INVALID_VIRTUAL_SERVERS;
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:124]: will we free all resources if we return here?
like if we have 5 servers and on the last one we have an issue will the other 4 close their fds?

File: src/class/ServerManager/ServerManager.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	return Poll::DONE;
+}
+
+Poll::ret_stt initSocketsHandler(Poll &p, int fd, int revents, Poll::Param &param)
+{
+	(void)p;
+	(void)fd;
+	(void)revents;
+	(void)param;
+	int client_fd = Server::acceptConnection(fd);
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:68]: `acceptConnection()` can throw except, who will catch it?

File: src/class/ServerManager/ServerManager.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	return Poll::DONE;
+}
+
+Poll::ret_stt client_read(Poll &p, int fd, int revents, Poll::Param &param)
+{
+	(void)p;
+	(void)fd;
+	(void)revents;
+	(void)param;
+	if (param.req.getRequestAttempts() >= HTTPRequest::MAX_CHUNK_ATTEMPTS)
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:35]: i don't think I understand the logic of CHUNK_ATTEMPTS.
is it about trying to read but receiving nothing?

File: src/class/ServerManager/ServerManager.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		return Poll::DONE;
+	try
+	{
+		param.req.recvRequest();
+		param.req.getBasicRequest().parseRaw();
+	}
+	catch (ABaseHTTPCall::Incomplete &e)
+	{
+		std::cerr << "Request is not finished [" << e.what() << "]\n";
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:44]: we need to `unParse();` in this case

File: src/class/HTTPRequest/HTTPRequest.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+/*
+** --------------------------------- METHODS ----------------------------------
+*/
+
+void HTTPRequest::recvRequest(void)
+{
+	int tmpRecvLen;
+	char tmpRaw[4096];
+
+	tmpRecvLen = recv(this->_clientFd, tmpRaw, sizeof(tmpRaw), MSG_DONTWAIT);
```
Comment (arikhativa) [inline src/class/HTTPRequest/HTTPRequest.cpp:57]: tmpRecvLen -> tmp_recv_len

File: src/class/HTTPRequest/HTTPRequest.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+/*
+** --------------------------------- METHODS ----------------------------------
+*/
+
+void HTTPRequest::recvRequest(void)
+{
+	int tmpRecvLen;
+	char tmpRaw[4096];
```
Comment (arikhativa) [inline src/class/HTTPRequest/HTTPRequest.cpp:55]: tmpRaw -> tmp_raw

File: src/class/HTTPRequest/HTTPRequest.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+	tmpRecvLen = recv(this->_clientFd, tmpRaw, sizeof(tmpRaw), MSG_DONTWAIT);
+	this->_requestAttempts++;
+	this->_rawRequest += tmpRaw;
+	if (tmpRecvLen <= -1)
```
Comment (arikhativa) [inline src/class/HTTPRequest/HTTPRequest.cpp:60]: i think there is a case of tmpRecvLen == 0 where the client closed the connection.
it is not an error, and we will need to remove the client from poll

File: src/class/ServerManager/ServerManager.cpp [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+		param.req.sendResponse();
+	}
+	catch (const std::exception &e)
+	{
+		std::cerr << e.what() << '\n';
+		return Poll::DONE;
+	}
+
+	if (param.req.getBytesSent() < param.req.getResponse().size() ||
+		param.req.getResponseAttempts() >= HTTPRequest::MAX_CHUNK_ATTEMPTS)
```
Comment (arikhativa) [inline src/class/ServerManager/ServerManager.cpp:23]: why param.req.getResponseAttempts() is here?

**Review by arikhativa — CHANGES_REQUESTED** [review]
remove all (void)

File: src/class/ServerManager/ServerManager.cpp (hunk 5 above)
Comment (dVaGaymer) [reply src/class/ServerManager/ServerManager.cpp]: I deleted it from now, it was something I thought nginx did while I was doing tests

**Review by arikhativa — APPROVED** [review]

**Pull request merged**